zebra-consensus/src/verify/redjubjub.rs

```diff
@@ -0,0 +1,149 @@
+use std::{
+    future::Future,
+    mem,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use rand::thread_rng;
+use redjubjub::*;
+use tokio::sync::broadcast::{channel, RecvError, Sender};
+use tower::Service;
+use tower_batch::BatchControl;
+
+/// RedJubjub signature verifier service
+pub struct RedJubjubVerifier<T: SigType> {
+    batch: batch::Verifier<T>,
+    // This uses a "broadcast" channel, which is an mpmc channel. Tokio also
+    // provides a spmc channel, "watch", but it only keeps the latest value, so
+    // using it would require thinking through whether it was possible for
+    // results from one batch to be mixed with another.
+    tx: Sender<Result<(), Error>>,
+}
+
+#[allow(clippy::new_without_default)]
+impl<T: SigType> RedJubjubVerifier<T> {
+    /// Create a new RedJubjubVerifier instance
+    pub fn new() -> Self {
+        let batch = batch::Verifier::<T>::default();
+        // XXX(hdevalence) what's a reasonable choice here?
+        let (tx, _) = channel(10);
+        Self { tx, batch }
+    }
+}
+
+/// Type alias to clarify that this batch::Item is a RedJubjubItem
+pub type RedJubjubItem<T> = batch::Item<T>;
+
+impl<'msg, T: SigType> Service<BatchControl<RedJubjubItem<T>>> for RedJubjubVerifier<T> {
```

Comment:
where is this `'msg` parameter used?

Reply:
Ah that's erratum from when we were batching messages in items, we pre-process messages now so we don't need this anymore.